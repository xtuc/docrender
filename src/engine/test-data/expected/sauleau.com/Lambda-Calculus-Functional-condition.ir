(
    content: [
        Clear,
        Title((
            level: 2,
            text: Text("Lambda Calculus"),
        ), []),
        Clear,
        Paragraph([
            InlineText([
                Text("If you have never heard about lambda calculus, here is a quick definition found on Wikipedia:"),
            ], true),
        ]),
        Clear,
        Paragraph([
            InlineText([
                Text("“Lambda calculus (also written as λ-calculus) is a formal system in mathematical logic for expressing computation based on function abstraction and application using variable binding and substitution.”"),
            ], true),
        ]),
        Clear,
        Paragraph([
            InlineText([
                Text("Basically, this is the syntax:"),
            ], true),
        ]),
        Clear,
        Paragraph([
            InlineText([
                Text("<function> ::= λ <variable-list> . <expression>"),
            ], true),
        ]),
        Clear,
        Paragraph([
            InlineText([
                Text("An expression is also a variable."),
            ], true),
        ]),
        Clear,
        Paragraph([
            InlineText([
                Text("For example, here is the "),
            ], false),
            InlineText([
                TextWithOptions("identity function from mathematics", (
                    link: Some("https://en.wikipedia.org/wiki/Identity_function"),
                    styles: [
                        ("text-decoration", "underline"),
                        ("color", "rgb(6.666667%, 33.333336%, 80%)"),
                    ],
                )),
            ], false),
            InlineText([
                Text(":"),
            ], true),
        ]),
        Clear,
        Paragraph([]),
        Code("lambda calculus", "identity = λx.x\n"),
        Clear,
        Paragraph([
            InlineText([
                Text("A more concrete example is its usage to represent types. In the following example:"),
            ], true),
        ]),
        Clear,
        Paragraph([]),
        Code("js", "function upper(str) {\n  return str.toUpperCase()\n}\n"),
        Clear,
        Paragraph([
            InlineText([
                Text("has the type:"),
            ], true),
        ]),
        Clear,
        Paragraph([]),
        Code("lambda calculus", "t\' = λ t . t\n\n// We can reduce it to infer the return type\nt\' = λ string . string\nt\' = string\n\n// The result type is `string`\n"),
        Clear,
        Paragraph([]),
        Title((
            level: 4,
            text: Text("Let\'s define booleans"),
        ), []),
        Clear,
        Code("lambda calculus", "true = λx.λy.x\nfalse = λx.λy.y    \n"),
        Clear,
        Title((
            level: 4,
            text: Text("Define the conditional operator"),
        ), []),
        Clear,
        Code("lambda calculus", "if = λb.λt.λf.b t f\n"),
        Clear,
        Title((
            level: 4,
            text: Text("Define the logical operators"),
        ), []),
        Clear,
        Code("lambda calculus", "and = λbb\'. if b b\' false\nor = λbb\'. if b true b\'\nnot = λb.if b false true\n"),
        Clear,
        Title((
            level: 3,
            text: Text("Example"),
        ), []),
        Clear,
        Paragraph([
            InlineText([
                Text("I decided to use OCaml for this example because it has partial application and a nice REPL."),
            ], true),
        ]),
        Clear,
        Paragraph([
            InlineText([
                Text("Since the conditional and logical operators are already in the language, I will prepend them with an underscore."),
            ], true),
        ]),
        Clear,
        Paragraph([]),
        Title((
            level: 4,
            text: Text("Translate definitions into OCaml"),
        ), []),
        Clear,
        Code("ocaml", "let _true x y = x;;\n(* - : \'a -> \'b -> \'a =  *)\n\nlet _false x y = y;;\n(* - : \'a -> \'b -> \'b =  *)\n"),
        Clear,
        Paragraph([
            InlineText([
                Text("Note that it\'s similar to how an identity function works."),
            ], true),
        ]),
        Clear,
        Paragraph([]),
        Code("ocaml", "let _if b t f = b t f;;\n(* - : (\'a -> \'b -> \'c) -> \'a -> \'b -> \'c =  *)\n"),
        Clear,
        Paragraph([
            InlineText([
                Text("For the sake of simplicity, I won\'t declare all the operators for now. We can already evaluate some simple conditions."),
            ], true),
        ]),
        Clear,
        Paragraph([]),
        Code("ocaml", "(* This is my condition, for now it\'s a constant `true` *)\nlet myCond = _true;;\n\n_if myCond 1 0;;\n(* - : int = 1 *)\n\n(* and if my condition is `false` *)\nlet myCond2 = _false;;\n\n_if myCond2 1 0;;\n(* - : int = 0 *)\n"),
        Clear,
        Title((
            level: 3,
            text: Text("Conclusion"),
        ), []),
        Clear,
        Paragraph([
            InlineText([
                Text("To be honest, I don\'t know why you would use this in real life code. There is no need to redefine the built-in condition mechanism."),
            ], true),
        ]),
        Clear,
        Paragraph([
            InlineText([
                Text("Thanks to "),
            ], false),
            InlineText([
                TextWithOptions("Danny Willems", (
                    link: Some("https://twitter.com/dwillems42"),
                    styles: [
                        ("text-decoration", "underline"),
                        ("color", "rgb(6.666667%, 33.333336%, 80%)"),
                    ],
                )),
            ], false),
            InlineText([
                Text(" for introducing me to this."),
            ], true),
        ]),
        Clear,
        Paragraph([]),
    ],
    styles: {
        "pbNORMAL_TEXT": [
            ("line-height", "1.7249999"),
        ],
    },
)