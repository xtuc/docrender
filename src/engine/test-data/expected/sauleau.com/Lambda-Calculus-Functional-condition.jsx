import { createElement, Fragment } from "https://unpkg.com/es-react";

function Heading({ style, id, level, children }) {
  const tag = "h" + level;
  return createElement(tag, { style, id }, children);
}

function Paragraph({ children }) {
  return createElement("p", {}, children);
}

function Code({ children }) {
  return createElement("code", {}, children);
}

function Image({ src }) {
  return createElement("img", { src });
}

function Link({ href }) {
  return createElement("a", { href });
}

function List({ children }) {
  return createElement("ul", {}, children);
}

function ListItem({ children }) {
  return createElement("li", {}, children);
}

export default function Doc({
  heading = Heading,
  paragraph = Paragraph,
  code = Code,
  image = Image,
  link = Link,
  list = List,
  listItem = ListItem,
} = {}) {
  return createElement(
    Fragment,
    {},
    createElement(heading, { level: 2, id: "pb-todo" }, `Lambda Calculus`),
    createElement(
      paragraph,
      {},
      `If you have never heard about lambda calculus, here is a quick definition found on Wikipedia:`,
      `
`,
    ),
    createElement(
      paragraph,
      {},
      `“Lambda calculus (also written as λ-calculus) is a formal system in mathematical logic for expressing computation based on function abstraction and application using variable binding and substitution.”`,
      `
`,
    ),
    createElement(
      paragraph,
      {},
      `Basically, this is the syntax:`,
      `
`,
    ),
    createElement(
      paragraph,
      {},
      `&lt;function&gt; ::= λ &lt;variable-list&gt; . &lt;expression&gt;`,
      `
`,
    ),
    createElement(
      paragraph,
      {},
      `An expression is also a variable.`,
      `
`,
    ),
    createElement(
      paragraph,
      {},
      `For example, here is the `,
      createElement(
        link,
        {
          style: {
            textDecoration: "underline",
            color: "rgb(6.666667%, 33.333336%, 80%)",
          },
          href: "https://en.wikipedia.org/wiki/Identity_function",
        },
        `identity function from mathematics`,
      ),
      `:`,
      `
`,
    ),
    createElement(paragraph, {}),
    createElement(
      code,
      { class: "language-lambda calculus" },
      `<codeblock-js>identity = λx.x
</codeblock-js>`,
    ),
    createElement(
      paragraph,
      {},
      `A more concrete example is its usage to represent types. In the following example:`,
      `
`,
    ),
    createElement(paragraph, {}),
    createElement(
      code,
      { class: "language-js" },
      `<codeblock-js>function upper(str) {
  return str.toUpperCase()
}
</codeblock-js>`,
    ),
    createElement(
      paragraph,
      {},
      `has the type:`,
      `
`,
    ),
    createElement(paragraph, {}),
    createElement(
      code,
      { class: "language-lambda calculus" },
      `<codeblock-js>t' = λ t . t

// We can reduce it to infer the return type
t' = λ string . string
t' = string

// The result type is \`string\`
</codeblock-js>`,
    ),
    createElement(paragraph, {}),
    createElement(
      heading,
      { level: 4, id: "pb-todo" },
      `Let's define booleans`,
    ),
    createElement(
      code,
      { class: "language-lambda calculus" },
      `<codeblock-js>true = λx.λy.x
false = λx.λy.y    
</codeblock-js>`,
    ),
    createElement(
      heading,
      { level: 4, id: "pb-todo" },
      `Define the conditional operator`,
    ),
    createElement(
      code,
      { class: "language-lambda calculus" },
      `<codeblock-js>if = λb.λt.λf.b t f
</codeblock-js>`,
    ),
    createElement(
      heading,
      { level: 4, id: "pb-todo" },
      `Define the logical operators`,
    ),
    createElement(
      code,
      { class: "language-lambda calculus" },
      `<codeblock-js>and = λbb'. if b b' false
or = λbb'. if b true b'
not = λb.if b false true
</codeblock-js>`,
    ),
    createElement(heading, { level: 3, id: "pb-todo" }, `Example`),
    createElement(
      paragraph,
      {},
      `I decided to use OCaml for this example because it has partial application and a nice REPL.`,
      `
`,
    ),
    createElement(
      paragraph,
      {},
      `Since the conditional and logical operators are already in the language, I will prepend them with an underscore.`,
      `
`,
    ),
    createElement(paragraph, {}),
    createElement(
      heading,
      { level: 4, id: "pb-todo" },
      `Translate definitions into OCaml`,
    ),
    createElement(
      code,
      { class: "language-ocaml" },
      `<codeblock-js>let _true x y = x;;
(* - : 'a -> 'b -> 'a =  *)

let _false x y = y;;
(* - : 'a -> 'b -> 'b =  *)
</codeblock-js>`,
    ),
    createElement(
      paragraph,
      {},
      `Note that it's similar to how an identity function works.`,
      `
`,
    ),
    createElement(paragraph, {}),
    createElement(
      code,
      { class: "language-ocaml" },
      `<codeblock-js>let _if b t f = b t f;;
(* - : ('a -> 'b -> 'c) -> 'a -> 'b -> 'c =  *)
</codeblock-js>`,
    ),
    createElement(
      paragraph,
      {},
      `For the sake of simplicity, I won't declare all the operators for now. We can already evaluate some simple conditions.`,
      `
`,
    ),
    createElement(paragraph, {}),
    createElement(
      code,
      { class: "language-ocaml" },
      `<codeblock-js>(* This is my condition, for now it's a constant \`true\` *)
let myCond = _true;;

_if myCond 1 0;;
(* - : int = 1 *)

(* and if my condition is \`false\` *)
let myCond2 = _false;;

_if myCond2 1 0;;
(* - : int = 0 *)
</codeblock-js>`,
    ),
    createElement(heading, { level: 3, id: "pb-todo" }, `Conclusion`),
    createElement(
      paragraph,
      {},
      `To be honest, I don't know why you would use this in real life code. There is no need to redefine the built-in condition mechanism.`,
      `
`,
    ),
    createElement(
      paragraph,
      {},
      `Thanks to `,
      createElement(
        link,
        {
          style: {
            textDecoration: "underline",
            color: "rgb(6.666667%, 33.333336%, 80%)",
          },
          href: "https://twitter.com/dwillems42",
        },
        `Danny Willems`,
      ),
      ` for introducing me to this.`,
      `
`,
    ),
    createElement(paragraph, {}),
  );
}
