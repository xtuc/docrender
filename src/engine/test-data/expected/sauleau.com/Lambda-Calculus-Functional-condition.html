<style>
  .pbNORMAL_TEXT {
    line-height: 1.7249999;
  }
</style>
<h2 id="pb-Lambda%20Calculus">Lambda Calculus</h2>
<p>
  If you have never heard about lambda calculus, here is a quick definition
  found on Wikipedia:
</p>
<p>
  “Lambda calculus (also written as λ-calculus) is a formal system in
  mathematical logic for expressing computation based on function abstraction
  and application using variable binding and substitution.”
</p>
<p>Basically, this is the syntax:</p>
<p>&lt;function&gt; ::= λ &lt;variable-list&gt; . &lt;expression&gt;</p>
<p>An expression is also a variable.</p>
<p>
  For example, here is the
  <a
    style="text-decoration: underline; color: rgb(6.666667%, 33.333336%, 80%)"
    href="https://en.wikipedia.org/wiki/Identity_function"
    >identity function from mathematics</a
  >:
</p>
<pre><code class="language-lambda calculus"><codeblock-js>identity = λx.x
</codeblock-js></code></pre>
<style>
  script[visible],
  style[visible],
  pre {
    display: block;
    white-space: pre;
    border: 1px solid #dedede;
    padding: 1em;
    background: #fafafa;
    overflow-x: auto;
    border-left: 0.4em solid cornflowerblue;
    tab-size: 2;
    color: #1a1a1a;
    line-height: 1.6;
  }

  code:not(pre code), /* output:not(code:has(output) output) */ {
    background: #f7f7f7;
    border: 1px solid rgb(0 0 0 / 0.2);
    padding: 0.1rem 0.3rem;
    margin: 0.1rem 0;
    border-radius: 0.2rem;
    display: inline-block;
  }
</style>
<p>
  A more concrete example is its usage to represent types. In the following
  example:
</p>
<pre><code class="language-js"><codeblock-js>function upper(str) {
  return str.toUpperCase()
}
</codeblock-js></code></pre>
<style>
  script[visible],
  style[visible],
  pre {
    display: block;
    white-space: pre;
    border: 1px solid #dedede;
    padding: 1em;
    background: #fafafa;
    overflow-x: auto;
    border-left: 0.4em solid cornflowerblue;
    tab-size: 2;
    color: #1a1a1a;
    line-height: 1.6;
  }

  code:not(pre code), /* output:not(code:has(output) output) */ {
    background: #f7f7f7;
    border: 1px solid rgb(0 0 0 / 0.2);
    padding: 0.1rem 0.3rem;
    margin: 0.1rem 0;
    border-radius: 0.2rem;
    display: inline-block;
  }
</style>
<p>has the type:</p>
<pre><code class="language-lambda calculus"><codeblock-js>t' = λ t . t

// We can reduce it to infer the return type
t' = λ string . string
t' = string

// The result type is `string`
</codeblock-js></code></pre>
<style>
  script[visible],
  style[visible],
  pre {
    display: block;
    white-space: pre;
    border: 1px solid #dedede;
    padding: 1em;
    background: #fafafa;
    overflow-x: auto;
    border-left: 0.4em solid cornflowerblue;
    tab-size: 2;
    color: #1a1a1a;
    line-height: 1.6;
  }

  code:not(pre code), /* output:not(code:has(output) output) */ {
    background: #f7f7f7;
    border: 1px solid rgb(0 0 0 / 0.2);
    padding: 0.1rem 0.3rem;
    margin: 0.1rem 0;
    border-radius: 0.2rem;
    display: inline-block;
  }
</style>
<h4 id="pb-Let%27s%20define%20booleans">Let's define booleans</h4>
<pre><code class="language-lambda calculus"><codeblock-js>true = λx.λy.x
false = λx.λy.y    
</codeblock-js></code></pre>
<style>
  script[visible],
  style[visible],
  pre {
    display: block;
    white-space: pre;
    border: 1px solid #dedede;
    padding: 1em;
    background: #fafafa;
    overflow-x: auto;
    border-left: 0.4em solid cornflowerblue;
    tab-size: 2;
    color: #1a1a1a;
    line-height: 1.6;
  }

  code:not(pre code), /* output:not(code:has(output) output) */ {
    background: #f7f7f7;
    border: 1px solid rgb(0 0 0 / 0.2);
    padding: 0.1rem 0.3rem;
    margin: 0.1rem 0;
    border-radius: 0.2rem;
    display: inline-block;
  }
</style>
<h4 id="pb-Define%20the%20conditional%20operator">
  Define the conditional operator
</h4>
<pre><code class="language-lambda calculus"><codeblock-js>if = λb.λt.λf.b t f
</codeblock-js></code></pre>
<style>
  script[visible],
  style[visible],
  pre {
    display: block;
    white-space: pre;
    border: 1px solid #dedede;
    padding: 1em;
    background: #fafafa;
    overflow-x: auto;
    border-left: 0.4em solid cornflowerblue;
    tab-size: 2;
    color: #1a1a1a;
    line-height: 1.6;
  }

  code:not(pre code), /* output:not(code:has(output) output) */ {
    background: #f7f7f7;
    border: 1px solid rgb(0 0 0 / 0.2);
    padding: 0.1rem 0.3rem;
    margin: 0.1rem 0;
    border-radius: 0.2rem;
    display: inline-block;
  }
</style>
<h4 id="pb-Define%20the%20logical%20operators">Define the logical operators</h4>
<pre><code class="language-lambda calculus"><codeblock-js>and = λbb'. if b b' false
or = λbb'. if b true b'
not = λb.if b false true
</codeblock-js></code></pre>
<style>
  script[visible],
  style[visible],
  pre {
    display: block;
    white-space: pre;
    border: 1px solid #dedede;
    padding: 1em;
    background: #fafafa;
    overflow-x: auto;
    border-left: 0.4em solid cornflowerblue;
    tab-size: 2;
    color: #1a1a1a;
    line-height: 1.6;
  }

  code:not(pre code), /* output:not(code:has(output) output) */ {
    background: #f7f7f7;
    border: 1px solid rgb(0 0 0 / 0.2);
    padding: 0.1rem 0.3rem;
    margin: 0.1rem 0;
    border-radius: 0.2rem;
    display: inline-block;
  }
</style>
<h3 id="pb-Example">Example</h3>
<p>
  I decided to use OCaml for this example because it has partial application and
  a nice REPL.
</p>
<p>
  Since the conditional and logical operators are already in the language, I
  will prepend them with an underscore.
</p>
<h4 id="pb-Translate%20definitions%20into%20OCaml">
  Translate definitions into OCaml
</h4>
<pre><code class="language-ocaml"><codeblock-js>let _true x y = x;;
(* - : 'a -> 'b -> 'a =  *)

let _false x y = y;;
(* - : 'a -> 'b -> 'b =  *)
</codeblock-js></code></pre>
<style>
  script[visible],
  style[visible],
  pre {
    display: block;
    white-space: pre;
    border: 1px solid #dedede;
    padding: 1em;
    background: #fafafa;
    overflow-x: auto;
    border-left: 0.4em solid cornflowerblue;
    tab-size: 2;
    color: #1a1a1a;
    line-height: 1.6;
  }

  code:not(pre code), /* output:not(code:has(output) output) */ {
    background: #f7f7f7;
    border: 1px solid rgb(0 0 0 / 0.2);
    padding: 0.1rem 0.3rem;
    margin: 0.1rem 0;
    border-radius: 0.2rem;
    display: inline-block;
  }
</style>
<p>Note that it's similar to how an identity function works.</p>
<pre><code class="language-ocaml"><codeblock-js>let _if b t f = b t f;;
(* - : ('a -> 'b -> 'c) -> 'a -> 'b -> 'c =  *)
</codeblock-js></code></pre>
<style>
  script[visible],
  style[visible],
  pre {
    display: block;
    white-space: pre;
    border: 1px solid #dedede;
    padding: 1em;
    background: #fafafa;
    overflow-x: auto;
    border-left: 0.4em solid cornflowerblue;
    tab-size: 2;
    color: #1a1a1a;
    line-height: 1.6;
  }

  code:not(pre code), /* output:not(code:has(output) output) */ {
    background: #f7f7f7;
    border: 1px solid rgb(0 0 0 / 0.2);
    padding: 0.1rem 0.3rem;
    margin: 0.1rem 0;
    border-radius: 0.2rem;
    display: inline-block;
  }
</style>
<p>
  For the sake of simplicity, I won't declare all the operators for now. We can
  already evaluate some simple conditions.
</p>
<pre><code class="language-ocaml"><codeblock-js>(* This is my condition, for now it's a constant `true` *)
let myCond = _true;;

_if myCond 1 0;;
(* - : int = 1 *)

(* and if my condition is `false` *)
let myCond2 = _false;;

_if myCond2 1 0;;
(* - : int = 0 *)
</codeblock-js></code></pre>
<style>
  script[visible],
  style[visible],
  pre {
    display: block;
    white-space: pre;
    border: 1px solid #dedede;
    padding: 1em;
    background: #fafafa;
    overflow-x: auto;
    border-left: 0.4em solid cornflowerblue;
    tab-size: 2;
    color: #1a1a1a;
    line-height: 1.6;
  }

  code:not(pre code), /* output:not(code:has(output) output) */ {
    background: #f7f7f7;
    border: 1px solid rgb(0 0 0 / 0.2);
    padding: 0.1rem 0.3rem;
    margin: 0.1rem 0;
    border-radius: 0.2rem;
    display: inline-block;
  }
</style>
<h3 id="pb-Conclusion">Conclusion</h3>
<p>
  To be honest, I don't know why you would use this in real life code. There is
  no need to redefine the built-in condition mechanism.
</p>
<p>
  Thanks to
  <a
    style="text-decoration: underline; color: rgb(6.666667%, 33.333336%, 80%)"
    href="https://twitter.com/dwillems42"
    >Danny Willems</a
  >
  for introducing me to this.
</p>
